docs/design/features/bridging-constraints.md

```diff
@@ -0,0 +1,110 @@
+    {
+        // The T in Do is unconstrainted, but DoConstrained requires it to be a struct
+        // The run time check above ensures this.
+        // This will not compile as C# today because the constraints don't match.
+        // Today, it will also result in a verification exception at runtime when compiling the Do method.
+        DoConstrained<T>();
+    }
+}
+```
+
+The purpose of this document is to sketch out a possible implementation on the runtime side that would allow above code to work.
+
+This feature doesn't assume any kind of relaxation of constraint checks outside method bodies. E.g. allowing to derive from `class Base<T> where T : IFoo { }` with `class Derived<T> : Base<T> { }` to load for `Derived<IFoo>` is not in scope.
+
+## Runtime impact
+
+There are two ways how constraints are validated within the runtime right now:
+
+* Validating they match on uninstantiated types: done by comparing constraints on generic type parameters - if type parameter T is used to substitute U, the constraints need to be satisfied.
+* Validating they match on concrete instantiation: for example `class Foo<T> where T: struct { }`: `Foo<int>` satisfies constraints, `Foo<object>` doesn't.
+
+Validating on uninstantiated forms is done as part of the old verification logic. The runtime doesn't do this consistently - for example, it is done in calls in IL, but not for ldtoken in IL.
+
+Validating instantiated forms is done always so that a `MethodTable` for an invalid type cannot even be created.
+
+From the runtime perspective, it would be desirable to maintain the invariant that `MethodTable` for a generic type instantiation that doesn't meet its own constraints can never be loaded - an attempt to do so would continue throwing an exception. This has implications on how method bodies that may be violating constraints in certain branches can be compiled. Consider:
+
+```csharp
+//
+// This is pseudo-C# just so that I don't have to write IL.
+// Not expected to compile as C# in this shape, but imagine the straightforward lowering to IL.
+//
+interface IInterface { }
+
+struct Constrained<T> where T : IInterface { }
+
+static void Do<T>()
+{
+    if (typeof(T).IsAssignableFrom(typeof(IInterface))
+    {
+        Constrained<T> someLocal = default;
+        try
+        {
+            someLocal.ToString();
+        }
+        catch (GenericException<Constrained<T>>) { }
+    }
+}
+```
+
+The requirement of never having an invalid type instantiation within the runtime type system means that if we're compiling `Do<int>` (where we cannot allow `Constrained<int>` to exist), the code generator cannot ask questions about all locals (IL locals are scoped to the method, not to a basic block), exception handling regions, or any IL within the `IsAssignableFrom` check. While it may be reasonable to expect the `IsAssignableFrom` check gets optimized out, it would likely not happen for unoptimized code. Locals and EH regions pose additional programs.
```

Comment:
> the code generator cannot ask questions about all locals

This is not limited to code generator. It extends to diagnostic tools or IL inspection APIs like https://learn.microsoft.com/dotnet/api/system.reflection.localvariableinfo.localtype

Reply:
Diagnostics tools are an important consideration here as this would introduce novel failure modes. Most modern debuggers have the "move IP to" gesture. Moving the IP within an invalid branch for the generic would require serious thought on the correct way to fail.